Doc/library/typing.rst

```diff
@@ -1230,6 +1232,83 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    Usage::
+
+        Ts = TypeVarTuple('Ts')
+        Shape = TypeVarTuple('Shape')
+
+    Vanilla type variables enable parameterization with a *fixed* number of
+    types. Type variable tuples, on the other hand, allow parameterization with
+    an *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        class Array(Generic[*Shape]):
+            pass
+
+        from typing import NewType
+        Height = NewType('Height', int)
+        Width = NewType('Width', int)
+
+        1d_array: Array[Height] = Array()  # This is fine
+
+    With one type argument, ``Array`` behaves as if ``Shape = (T,)``, where
+    ``T`` is a type variable. That means ``Array`` behaves as if we'd written
+    ``Generic[T]``. Note the use of the unpacking operator ``*`` to bring ``T``
+    out of the tuple and directly into the square brackets.
+
+    ::
+
+        2d_array: Array[Height, Width] = Array()  # This is also fine!
+
+    With two arguments, it's as if ``Shape = (T1, T2)``, where ``T1`` and ``T2``
+    are type variables; as if ``Array`` was defined with ``Generic[T1, T2]``.
+
+    Type variable tuples can be used any place a vanilla type variable can.
+    However, note that they must always be used in combination with the
+    unpacking operator ``*``. This helps distinguish type variable tuples from
+    normal type variables::
+
+        def returns_tuple() -> Ts:  # Not valid
+          return ('spam', 42)
+
+        def returns_tuple() -> tuple[*Ts]:  # This is fine :)
+          return ('spam', 42)
+
+    Type variable tuples can be happily combined with vanilla type variables::
+
+        DType = TypeVar('DType')
+
+        class Array(Generic[DType, *Shape]):  # This is fine
+            pass
+
+        1d_float_array: Array[float, Height] = Array()     # Totally fine
+        2d_int_array: Array[int, Height, Width] = Array()  # Yup, fine too
+
+    However, note that at most one type variable tuple may appear in a single
+    list of type arguments or type parameters::
+
+        class Array(Generic[*Shape, *Shape]):  # Not valid
+            pass
```

Comment:
You are carefully not bringing up `Generic[*Shape, T]`. Is it valid?

Reply:
We show `Generic[DType, *Shape]` - is that not enough? Or is the point that we don't show e.g. `Generic[*Shape, DType]`? We could explicitly call it out as also valid, but I don't think the average reader would have any reason to think it wasn't possible.

Reply:
It's worth calling out actually, because in a function definition `def f(*args, x)` has a different meaning than `def f(x, *args)`.

Reply:
Ah, good point. Done.